crates/client-api-messages/src/ws.rs

```diff
@@ -0,0 +1,266 @@
+    /// Register SQL queries on which to receive updates.
+    Subscribe(Subscribe),
+    /// Send a one-off SQL query without establishing a subscription.
+    OneOffQuery(OneOffQuery),
+}
+
+impl<Args> ClientMessage<Args> {
+    pub fn map_args<Args2>(self, f: impl FnOnce(Args) -> Args2) -> ClientMessage<Args2> {
+        match self {
+            ClientMessage::CallReducer(CallReducer {
+                reducer,
+                args,
+                request_id,
+            }) => ClientMessage::CallReducer(CallReducer {
+                reducer,
+                args: f(args),
+                request_id,
+            }),
+            ClientMessage::Subscribe(x) => ClientMessage::Subscribe(x),
+            ClientMessage::OneOffQuery(x) => ClientMessage::OneOffQuery(x),
+        }
+    }
+}
+
+#[derive(Deserialize, Serialize, serde::Deserialize)]
+pub struct CallReducer<Args> {
+    /// The name or id of the reducer to call.
+    pub reducer: ReducerId,
+    /// The arguments to the reducer, encoded as ????.
+    pub args: Args,
+    /// An identifier for a client request
+    pub request_id: u32,
+}
+
+/// A specification of the reducer to call for [`CallReducer`].
+#[derive(Deserialize, Serialize, serde::Deserialize)]
+pub enum ReducerId {
+    /// A reducer specified by its name.
+    Name(String),
+    // /// A reducer specified by its numerical ID.
+    // Id(u32),
+}
+
+/// Sent by client to database to register a set of queries, about which the client will
+/// receive `TransactionUpdate`s.
+///
+/// After issuing a `Subscribe` message, the client will receive a single
+/// `SubscriptionUpdate` message containing every current row of every table which matches
+/// the subscribed queries. Then, after each reducer run which updates one or more
+/// subscribed rows, the client will receive a `TransactionUpdate` containing the updates.
+///
+/// A `Subscribe` message sets or replaces the entire set of queries to which the client
+/// is subscribed. If the client is previously subscribed to some set of queries `A`, and
+/// then sends a `Subscribe` message to subscribe to a set `B`, afterwards, the client
+/// will be subscribed to `B` but not `A`. In this case, the client will receive a
+/// `SubscriptionUpdate` containing every existing row that matches `B`, even if some were
+/// already in `A`.
+#[derive(Deserialize, Serialize, serde::Deserialize)]
+pub struct Subscribe {
+    /// A sequence of SQL queries.
+    pub query_strings: Vec<String>,
+    pub request_id: u32,
+}
+
+/// A one-off query submission.
+///
+/// Query should be a "SELECT * FROM Table WHERE ...". Other types of queries will be rejected.
+/// Multiple such semicolon-delimited queries are allowed.
+///
+/// One-off queries are identified by a client-generated messageID.
+/// To avoid data leaks, the server will NOT cache responses to messages based on UUID!
+/// It also will not check for duplicate IDs. They are just a way to match responses to messages.
```

Comment:
Doc comments should be moved to fields.

Reply:
Ya - I tried to do that for the most part; most/all of the comments are copied from the protobuf, which tended to put field docs on the outer type.